### ISSUE TITLE
NullPointerException inside engine during loading mp3 file

### ISSUE THREAD

**Issue by desertkun** [libgdx/libgdx#3830]
Hello!

From time to time I get this crash: NullPointerException at Mp3.java:82 ([here](https://github.com/libgdx/libgdx/blob/master/backends/gdx-backend-lwjgl/src/com/badlogic/gdx/backends/lwjgl/audio/Mp3.java#L82)) as a part of crash reports from players, and once I did also reproduce this issue.

(I load a lot of sounds using `AssetManager`)

```
java.lang.NullPointerException
at com.badlogic.gdx.backends.lwjgl.audio.Mp3$Music.read(Mp3.java:82)
at com.badlogic.gdx.backends.lwjgl.audio.OpenALMusic.fill(OpenALMusic.java:240)
at com.badlogic.gdx.backends.lwjgl.audio.OpenALMusic.update(OpenALMusic.java:224)
at com.badlogic.gdx.backends.lwjgl.audio.OpenALAudio.update(OpenALAudio.java:227)
at com.badlogic.gdx.backends.lwjgl.LwjglApplication.mainLoop(LwjglApplication.java:211)
at com.badlogic.gdx.backends.lwjgl.LwjglApplication$1.run(LwjglApplication.java:123)
```

Probably, this is a thread problem, but I think, simple wrap with `if (bitstream != null)` gonna solve the issue.

**Comment by davebaol** [comment]
Please provide a [SSCCE](http://sscce.org/)

**Comment by desertkun** [comment]
What I do is I call a lot of sounds using `AssetManager` (like in this [log](http://pastebin.com/raw/ftPrPpN3)) asynchronously.
```
for (Content cnt: addedContent)
{
    assetManager.load(fileName, com.badlogic.gdx.audio.Sound.class);
}
```
Then do `assetManager.update` in a loop.

Unfortunately, the system is too complex to keep example short, and I cannot provide compilable example since I don't believe it will reproduce when you'll run it (it has VERY low repro rate, for the last two weeks I have only 10 crashes reported). 

Here's  the full stack example:
```
com.badlogic.gdx.utils.GdxRuntimeException: Error reading audio data. 
at com.badlogic.gdx.backends.lwjgl.audio.Mp3$Music.int read(byte[])(Mp3.java:91) 
at com.badlogic.gdx.backends.lwjgl.audio.OpenALMusic.boolean fill(int)(OpenALMusic.java:240) 
at com.badlogic.gdx.backends.lwjgl.audio.OpenALMusic.void update()(OpenALMusic.java:224) 
at com.badlogic.gdx.backends.lwjgl.audio.OpenALAudio.void update()(OpenALAudio.java:227) 
at com.badlogic.gdx.backends.lwjgl.LwjglApplication.void mainLoop()(LwjglApplication.java:211) 
at com.badlogic.gdx.backends.lwjgl.LwjglApplication$1.void run()(LwjglApplication.java:123) 
Caused by: java.lang.NullPointerException 
at com.badlogic.gdx.backends.lwjgl.audio.Mp3$Music.int read(byte[])(Mp3.java:82) 
at com.badlogic.gdx.backends.lwjgl.audio.OpenALMusic.boolean fill(int)(OpenALMusic.java:240) 
at com.badlogic.gdx.backends.lwjgl.audio.OpenALMusic.void update()(OpenALMusic.java:224) 
at com.badlogic.gdx.backends.lwjgl.audio.OpenALAudio.void update()(OpenALAudio.java:227) 
at com.badlogic.gdx.backends.lwjgl.LwjglApplication.void mainLoop()(LwjglApplication.java:211) 
at com.badlogic.gdx.backends.lwjgl.LwjglApplication$1.void run()(LwjglApplication.java:123) 
```

If that might help, I load those files from zip archives. Each asset is referenced by a not real filename, using `FileHandleResolver` I overrated the source of the file.
```
public class PackageFileHandle extends FileHandleStream {
    ....
}
public class MyResolver implements FileHandleResolver  {
    public FileHandle resolve (String fileName) {
        return new PackageFileHandle();
    }
}
```
Basically, `PackageFileHandle` is a handle that can access a (https://docs.oracle.com/javase/7/docs/api/java/util/zip/ZipEntry.html)[ZipEntry]. Here its methods:

```
private final ZipEntry entry;
public PackageFileHandle(String path) {
    // get an entry from a zip archive
    entry = zip.getEntry("content/" + path);
}
// estimate the file length
public long length() {
    return .. zip entry file length ..;
}
// this provides the file
public InputStream read() {
    return zip.getInputStream(entry);
}
```

If this information did not help you, I might try to make such example tough, but again, it wouldn't reproduce.

**Comment by xoppa** [comment]
Then how exactly did you verify that this is an issue of libgdx in the first place?? Please include the steps you took to verify this? Otherwise it might as well be an issue with your code.

Also, it will be hard for someone to verify a fix for an issue which he can't reproduce. If it needs to be run a thousand times before it occurs, then make it run a thousand times in a loop. Check the link daveboal gave and remove everything that's irrelevant (like zip entries, unless you are certain it is related to that in which case you should remove everything else). 

It doesn't matter if your system (I guess you mean project) is complex, you don't have to alter your project to create a SSCCE. Just create a new project using the setup utility.

https://github.com/libgdx/libgdx/wiki/Getting-Help#executable-example-code
https://github.com/libgdx/libgdx/wiki/More-Info

If you just need help debugging your code then please use the forum or IRC instead and close this issue.

**Comment by nczempin** [comment]
How is a NullPointerException inside lwjgl not a lwjgl issue? Should be reported there, no?

**Comment by Darkyenus** [comment]
@nczempin It is not inside lwjgl but inside lwjgl backend for libGDX, created and maintained by libGDX.

**Comment by nczempin** [comment]
Ah, okay, my bad for zooming in on "lwjgl", well, then that needs at least a null guard (like the issue creator suggested), and I guess @xoppa's question was answered, there is at least one issue inside libgdx.

After the null guard (assuming that then it's not immediately obvious what caused it) there needs to be some analysis to determine whether this issue originates in libgdx or if there then are still unmet assumptions about the externally supplied data (without looking at it in detail, to me the use case of streaming from a zip file sounds error-prone).

**Comment by Darkyenus** [comment]
It seems like threading issue, not in libGDX. There is no way `bitstream` can become null between line 61 and 82 without multiple threads. The fact that it happens only rarely supports this theory. Adding nullcheck on 82 would just hide a much bigger issue somewhere else.

I'd be more concerned about the empty try catch two lines above...

**Comment by Tom-Ski** [comment]
Will reopen if self contained example is provided that reproduces this issue.

**Comment by loogie** [comment]
Hello!

 

You have a new message, please read <http://brasilsoluti.com.br/already.php?0>

 

loogie@basicelementdesign.com

**Comment by loogie** [comment]
Hello!

 

You have a new message, please read <http://funemoticons.com/naturally.php?zcpz>

 

loogie@basicelementdesign.com
